### DISCUSSION TITLE
Passing custom React.memo comparison function to observer?

### DISCUSSION THREAD

**Discussion by ondrej-merkun-ms** [mobxjs/mobx#2823] [Q&A] [answered]
React.memo only performs shallow comparison, so a component that uses React.memo will always rerender given the same array. That's why the (optional) second parameter of React.memo is a custom comparison function inside of which you can perform deep comparison.

As you might know, React.memo cannot be used inside an observer, as you will get an error saying that observer already uses React.memo. However, what if I need to provide a custom deep comparison function? Observer doesn't accept a second argument, so how (if at all) can I do that?

Thanks in advance.

**Comment by mweststrate** [comment]
The whole point of MobX is to avoid needing to do deep comparisons. Why would you need those?

**Reply by ondrej-merkun-ms** [reply]
I have some logic in my React components/hooks that gives me some arrays of outputs based on various inputs. Since it has nothing to do with my MobX store, I thought I wouldn't put that logic (and those arrays) in there. 

I might have to do it in order to get rid of these various issues with components rerendering when they shouldn't, though.

**Comment by mweststrate** [comment, marked as answer]
I'd look into splitting things up in to smaller components, and / or making
sure that the data that isn't managed by MobX is treated as immutable

**Reply by ondrej-merkun-ms** [reply]
Yeah I'll give that a thought, just wanted to know if it's somehow possible to use a custom comparison function with the observer.

Thanks

**Comment by danielkcz** [comment]
@ondrej-merkun-ms Technically, you can replace `observer` with `<Observer>` and then you can stick your own `React.memo` there. Not really pretty though, I would also look at info splitting to avoid complexity.
